kolibri/core/public/api.py

```diff
@@ -146,3 +156,90 @@ def get_public_file_checksums(request, version):
         json.dumps({"id": error_constants.NOT_FOUND, "metadata": {"view": ""}}),
         content_type="application/json",
     )
+
+
+def position_in_queue(id):
+    qs = SyncQueue.objects.all().order_by("datetime").values("id")
+    pos = 0  # in case the queue is empty
+    for pos, value in enumerate(qs):
+        if value["id"] == id:
+            break
+    return pos
+
+
+class SyncQueueViewSet(viewsets.ViewSet):
+    def list(self, request):
+        """Returns length of the queue for each of the available facilities"""
+        SyncQueue.clean_stale()  # first, ensure not expired devices are in the queue
+        facilities = Facility.objects.all()
+        queue = {}
+        for facility in facilities:
+            queue[facility.id] = SyncQueue.objects.filter(
+                user__facility=facility
+            ).count()
+        return Response(queue)
+
+    def check_queue(self, pk=None):
+        current_transfers = TransferSession.objects.filter(active=True).count()
+        if current_transfers <= MAX_CONCURRENT_SYNCS:
+            allow_sync = True
+            data = {"action": SYNC}
+        else:
+            polling = current_transfers + SyncQueue.objects.all().count()
+            if pk is not None:
+                # if updating the element let's assign
+                # its time depending on its position in the queue
+                polling = current_transfers + position_in_queue(pk)
+            data = {
+                "action": QUEUED,
+                "keep_alive": polling,
+            }
+            allow_sync = False
+        return (allow_sync, data)
+
+    def create(self, request):
+        SyncQueue.clean_stale()  # first, ensure not expired devices are in the queue
+        device_info = get_device_info()
```

Comment:
Can also use `get_device_setting` utility here just to get the single value.

https://github.com/learningequality/kolibri/blob/814e00f9400a810986c76a47203c094fd6f3f42c/kolibri/core/device/utils.py#L22 (with a default of `False`)

Like it is used in the device info function: https://github.com/learningequality/kolibri/blob/develop/kolibri/core/public/utils.py#L16

Reply:
I thought using get_device_info would be preferable as it considers the case when the device is not provisioned. If that can't ever happen, perfect, I'll change it.

Reply:
Can just use the `default` kwarg to return False in that case - means we're only grabbing the info we need.

Reply:
Done